### PR TITLE
Don't retry messages unnecessarily

### PR DESCRIPTION
**Pull request by ulope**
## Description

Fixes: #4111
Fixes: #4600

Changes the order of constructing the messages to be sent and their
removal from the RetryQueue. This prevents unnecessary duplicate sending.
Also adds a test for this behaviour.

Since this was touching the same places this additionally implements killing idle RetryQueues after a timeout (#4600).

### CODE REVIEW

File: raiden/network/transport/matrix/transport.py [hunk 1]
```diff
@@ -238,6 +243,10 @@ def message_is_in_queue(data: _RetryQueue._MessageData) -> bool:
                     message=msg_data.message,
                     reason="Message was removed from queue",
                 )
+            else:
+                if next(msg_data.expiration_generator):
```
Comment (hackaugusto) [inline raiden/network/transport/matrix/transport.py]: How does this fix the problem, to me it looks that the if was just moved around

Comment (hackaugusto) [reply raiden/network/transport/matrix/transport.py]: nvm, the `for` loop is over `_message_queue` and not `message_texts`, so it was ignoring the expiration generator.

Comment (hackaugusto) [reply raiden/network/transport/matrix/transport.py]: but wait, the `_send_raw` is only called with data from `message_texts`, so it doesn't matter that it is iterating over all the messages.

Comment (ulope) [reply raiden/network/transport/matrix/transport.py]: The problem was that before the `message_texts` was constructed unconditionally before this loop (without paying attention if the messages were still in the raiden queues).

File: raiden/tests/mocked/test_matrix_transport.py [hunk 2]
```diff
@@ -161,13 +197,70 @@ def test_sending_nonstring_body(  # pylint: disable=unused-argument
...
+    with retry_queue._lock:
+        retry_queue._check_and_send()
+
+    assert len(mock_matrix.sent_messages) == 1
+    assert (factories.HOP1, serialized_message) in mock_matrix.sent_messages
+
+    mock_matrix._queueids_to_queues[queue_identifier].clear()
+
+    # The retry interval for this test is set to 0.01 (see parametrize). Make sure we're past this.
+    gevent.sleep(0.05)
```
Comment (hackaugusto) [inline raiden/tests/mocked/test_matrix_transport.py]: please use the fixture here, it will prevent errors if in the future we just change the fixture without reading all the tests.

File: raiden/tests/mocked/test_matrix_transport.py [hunk 3]
```diff
@@ -161,13 +197,70 @@ def test_sending_nonstring_body(  # pylint: disable=unused-argument
...
     invalid_message = '{"_type": "NonExistentMessage", "is": 3, "not_valid": 5}'
-    room, event = make_message(overwrite_data=invalid_message)
+    room, event = make_message_text(overwrite_data=invalid_message)
     assert not mock_matrix._handle_sync_messages([(room, [event])])
+
+
+@pytest.mark.parametrize("retry_interval", [0.01])
+def test_retry_queue_send_only_once(mock_matrix, record_sent_messages):
+    # Pretend the Transport greenlet is running
+    mock_matrix.greenlet = True
```
Comment (hackaugusto) [inline raiden/tests/mocked/test_matrix_transport.py]: ~Can we write this test without a mock of the transport?~ humm, not a the moment, we have to fix that.

File: raiden/tests/mocked/test_matrix_transport.py [hunk 4]
```diff
@@ -161,13 +197,70 @@ def test_sending_nonstring_body(  # pylint: disable=unused-argument
...
     mock_matrix, skip_userid_validation
 ):
     invalid_message = '{"_type": "NonExistentMessage", "is": 3, "not_valid": 5}'
-    room, event = make_message(overwrite_data=invalid_message)
+    room, event = make_message_text(overwrite_data=invalid_message)
     assert not mock_matrix._handle_sync_messages([(room, [event])])
+
+
+@pytest.mark.parametrize("retry_interval", [0.01])
+def test_retry_queue_send_only_once(mock_matrix, record_sent_messages):
```
Comment (hackaugusto) [inline raiden/tests/mocked/test_matrix_transport.py]: I think I understand the problem now:

- `_check_and_send` is called when a new message is added to the queue
- this will send the message once
- The `RetryQueue.run` will wait on `_notify_event`, once the timeout happens it will retry by calling `_check_and_send`.
- Since a timeout happened, a retry should be done, and the message is added to the list of "messages to resent" before checking if the message was removed. Therefore, it will send delivered messages again.

Please, improve the name of this test and add a docstring. Suggestions

Test name: `test_retry_queue_must_not_resent_delivered_messages`
Docstring:
```
RetryQueue used to retry messages even if the message was removed from the queue.
This regression test makes sure that if a message has been delivered, it will not be retried. 
```

File: raiden/network/transport/matrix/transport.py [hunk 5]
```diff
@@ -200,28 +205,28 @@ def _check_and_send(self) -> None:
...
+            #   - Those are retried according to their retry generator as long as they haven't been
+            #     removed from the Raiden queue
             remove = False
             if isinstance(msg_data.message, (Delivered, Ping, Pong)):
                 # e.g. Delivered, send only once and then clear
                 # TODO: Is this correct? Will a missed Delivered be 'fixed' by the
                 #       later `Processed` message?
                 remove = True
+                message_texts.append(msg_data.text)
             elif msg_data.queue_identifier not in self.transport._queueids_to_queues:
```
Comment (hackaugusto) [inline raiden/network/transport/matrix/transport.py]: can you please simplify this code, this `if` and the next are doing the same thing, pretty much.

File: raiden/network/transport/matrix/transport.py [hunk 6]
```diff
@@ -238,6 +243,10 @@ def message_is_in_queue(data: _RetryQueue._MessageData) -> bool:
                     message=msg_data.message,
                     reason="Message was removed from queue",
                 )
+            else:
+                if next(msg_data.expiration_generator):
+                    # Only (re-)send this message if the retry period has elapsed
```
Comment (hackaugusto) [inline raiden/network/transport/matrix/transport.py]: could you please add `and the message has not been delivered` or something in this form here?

File: raiden/network/transport/matrix/transport.py [hunk 7]
```diff
@@ -256,16 +265,27 @@ def _run(self) -> None:  # type: ignore
...
                 self._notify_event.clear()
                 if self._message_queue:
+                    self._idle_since = 0
                     self._check_and_send()
+                else:
+                    self._idle_since += 1
+
+            if self.is_idle:
+                self.log.debug("Exiting idle RetryQueue", queue=self)
+                return
```
Comment (hackaugusto) [inline raiden/network/transport/matrix/transport.py]: this doesn't look right to me, as far as I know there is no place to *restart* the RetryQueue, if our partner node goes offline, the number of retries is reached, the queue is considered idle and then exits, once the partner comes back online we won't be able to communicate with it.

File: raiden/tests/mocked/test_matrix_transport.py (hunk 3 above)
Comment (ulope) [reply raiden/tests/mocked/test_matrix_transport.py]: Yeah, the level of interdependency between the Transport and RetryQueue is pretty bad. 
#5607 summarizes that nicely, thanks.

File: raiden/tests/mocked/test_matrix_transport.py (hunk 4 above)
Comment (ulope) [reply raiden/tests/mocked/test_matrix_transport.py]: > Therefore, it will send delivered messages again.

Not `Delivered` but every retryable message can potentially be resent (it they're retry timeout has expired).

File: raiden/network/transport/matrix/transport.py (hunk 7 above)
Comment (ulope) [reply raiden/network/transport/matrix/transport.py]: They are never restarted. They only become idle if their message queue is empty for the specified amount of iterations. 
If then a message needs to be sent to the same peer `MatrixTransport._get_retrier()` will instantiate a new object.
I'll clarify the comments and also add a test for the negative case.

Comment (hackaugusto) [reply raiden/network/transport/matrix/transport.py]: Ahh, I see, I missed the check for `_message_queue`. Why not clearing as soon as it is empty?

On a side note: What about synchronization, this should only be done if there is no thread concurrently adding a message, is the `self._lock` being acquired in the `MatrixTransport`? (Humm, I don't really like this, having a lock being acquired outside of the owner class is usually a pain to understand)

Comment (ulope) [reply raiden/network/transport/matrix/transport.py]: The lock is acquired in `RetryQueue._run` (and I added an assert in this PR inside `_check_and_send` to ensure that's the case).

`MatrixTransport._get_retrier()` is, in the case we're talking about, just a dict lookup.
And in the two places it's called, the `enqueue()` method is called immediately afterwards. Therefore there can't be a gevent context switch.

Of course that could change. But yeah holding the lock from the outside seems wrong...

Comment (hackaugusto) [reply raiden/network/transport/matrix/transport.py]: Not really, `_get_retrier` will also initialize a new greenlet, meaning it will be schedule and I guess it could context-switch.

Did you measure the impact of this change? Is it really necessary? I have the impression this will introduce hard to debug bugs.

Another approach is to simply wait on the `_notify_event` *without* a timeout if there are no messages in the queue, then there will be no race conditions.

**Comment by ulope** [comment]
Thanks @hackaugusto.

Issue to track the idle handling: #5632
